lib/evmone/analysis.cpp

```diff
@@ -71,6 +71,28 @@ code_analysis analyze(evmc_revision rev, const uint8_t* code, size_t code_size)
 
         block.gas_cost += opcode_info.gas_cost;
 
+        // Optimize away heavy operations discarded by POP
+        if (code_pos != code_end && *code_pos == OP_POP && opcode_info.stack_req >= 1 &&
+            opcode_info.stack_change == 0)
+        {
+            bool read_only_and_const_cost = false;
+            switch (opcode)
+            {
+            case OP_BALANCE:
+            case OP_EXTCODESIZE:
+            case OP_EXTCODEHASH:
```

Comment:
Removing these won't mess with EIP-161 rules of "touching"?

Reply:
It will not. Per [EIP-106](https://eips.ethereum.org/EIPS/eip-161), an account is considered to be touched when it's involved in a potential value transfer (SUICIDE beneficiary, CALL* value sender or recipient, miner rewards, CREATE(2) endowment).

Reply:
While it is not the case yet, there are proposals now, which would mean this optimisation would need to be potentially disabled in a later hard fork. I think leaving a comment here for implications would be nice.

I'm still not sure SLOAD is not affected by the net gas metering EIP, haven't checked.

Reply:
@yperbasis did you progress through istanbul with this change yet? Does any client implement such optimistation?

Reply:
I've added more comments.

Only the cost of SSTORE, not SSLOAD is affected by [EIP-2200](https://eips.ethereum.org/EIPS/eip-2200).

We're using this optimization in a new C++ Ethereum client [Silkworm](https://github.com/torquem-ch/silkworm) (an evolution of Turbo-Geth). Istanbul blocks work fine with this change.

Reply:
Btw, if you want more inspiration for optimisations, check out Solidity's ruleset together with the `libevmasm/SemanticInformation` file :wink: